testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/ScreenshotUtil.java

```diff
@@ -0,0 +1,257 @@
+  private XmlSerializer mXmlSerializer;
+  private AlbumImpl album;
+  private OutputStream mOutputStream;
+
+  private static ScreenshotUtil instance;
+  private static int BUFFER_SIZE = 1 << 16; // 64K
+
+  private static ScreenshotUtil getInstance() {
+    synchronized (ScreenshotUtil.class) {
+      if (instance == null) {
+        instance = new ScreenshotUtil();
+      }
+      return instance;
+    }
+  }
+
+  /** Starts the album, which contains the screenshots in a zip file, and a metadata.xml file. */
+  void init() {
+    if (mXmlSerializer != null) {
+      return;
+    }
+    album = AlbumImpl.create(Registry.getRegistry().instrumentation.getContext(), "default");
+    album.cleanup();
+
+    mXmlSerializer = Xml.newSerializer();
+    try {
+      mOutputStream =
+          new BufferedOutputStream(new FileOutputStream(album.getMetadataFile()), BUFFER_SIZE);
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    try {
+      mXmlSerializer.setOutput(mOutputStream, "utf-8");
+      mXmlSerializer.startDocument("utf-8", null);
+      // Start tag <screenshots>.
+      mXmlSerializer.startTag(null, "screenshots");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  void writeText(String tagName, String value) throws IOException {
+    if (mXmlSerializer == null) {
+      throw new RuntimeException("ScreenshotUtil must be initialized.");
+    }
+    mXmlSerializer.startTag(null, tagName);
+    mXmlSerializer.text(value);
+    mXmlSerializer.endTag(null, tagName);
+  }
+
+  void writeBitmap(Bitmap bitmap, String name, String testClass, String testName)
+      throws IOException {
+    if (mXmlSerializer == null) {
+      throw new RuntimeException("ScreenshotUtil must be initialized.");
+    }
+    album.writeBitmap(name, 0, 0, bitmap);
+
+    mXmlSerializer.startTag(null, "screenshot");
+    writeText("name", name);
+    writeText("test_class", testClass);
+    writeText("test_name", testName);
+    writeText("tile_width", "1");
+    writeText("tile_height", "1");
+    mXmlSerializer.endTag(null, "screenshot");
+  }
+
+  /** Finishes metadata.xml. */
+  void flush() {
+    if (mXmlSerializer == null) {
+      throw new RuntimeException("ScreenshotUtil must be initialized.");
+    }
+    try {
+      // End tag </screenshots>
+      mXmlSerializer.endTag(null, "screenshots");
+      mXmlSerializer.endDocument();
+      mXmlSerializer.flush();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    try {
+      mOutputStream.close();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    album.flush();
+
+    mXmlSerializer = null;
+    mOutputStream = null;
+    album = null;
+  }
+
+  private static int getStatusBarHeight() {
+    final Context context = InstrumentationRegistry.getTargetContext();
+    final int resourceId =
+        context.getResources().getIdentifier("status_bar_height", "dimen", "android");
```

Comment:
are these magic strings? Can we add some code comments? So the next maintainer has some breadcrumbs in case anything needs to be changed.

Reply:
Done.